src/stores/widgets/StopGapWidgetDriver.ts

```diff
@@ -131,13 +131,11 @@ export class StopGapWidgetDriver extends WidgetDriver {
             }
         }
 
-        const allAllowed = new Set(iterableUnion(allowedSoFar, requested));
```

Comment:
what was the exact issue with the union? If the widget is approved for `[A, B, C, D]` and the widget requested `[B, C, E]`, then the widget should only receive `[B, C]` (the common permissions between the two). It should *not* be approved for things it didn't request, and obviously not for things it wasn't approved for in the first place.

Reply:
Oh, we named it `union` when it was doing a `merge`? That's backwards...

Reply:
We should restore this as an intersection, I guess.

Reply:
> what was the exact issue with the union? If the widget is approved for `[A, B, C, D]` and the widget requested `[B, C, E]`, then the widget should only receive `[B, C]` (the common permissions between the two). It should _not_ be approved for things it didn't request, and obviously not for things it wasn't approved for in the first place.

But when the user is asked and approves `E, it should be included in the list. The original issue was that this function not only just returned `[B, C]`, but it also _stored_ `[B, C]` as approved capabilities in the local storage, practically forgetting the prior approved `[A, D]` and requesting these permissions on every reload. When I understand https://github.com/matrix-org/matrix-doc/pull/2974 correctly, the re-exchange should be additive and it should indeed respond with `[A, B, C, D, E]` (iff the user approved `E`).

> Oh, we named it union when it was doing a merge? That's backwards...

I'm not sure what you mean by "doing a merge". For me that sounds suspiciously like a union of two sets 🤔. Which would make sense to be stored in the local storage (of course not with the requested but only with the approved ones). But given the "to send over the total list of requested and approved capabilities throughout the session" from the MSC, it shouldn't be necessary to calculate an intersection between the requested and approved sets(?).

We also might want to increase the test coverage of this code in the future, given its sensitivity (even though the widget API is still in a quite early state).

Reply:
If the widget didn't ask for `E` then it categorically should not be approved to use that capability. Whatever the user last stored should then be persisted as approved, as they've implicitly denied the capabilities not requested.

MSC2974 is additive, though the function here is not. It's a somewhat known limitation that requires breaking the API surface of the widget-api, not fixing here. If the widget never requested `A` in the widget session though, it shouldn't be approved to use it.

tldr: it's supposed to be an intersection, even if that introduces/maintains bugs at the moment.

Reply:
> tldr: it's supposed to be an intersection, even if that introduces/maintains bugs at the moment.

Ok that makes sense to me and this is probably why `// TODO: Do something when the widget requests new capabilities not yet asked for` is there, which I didn't properly understand before. So the unexpected behavior we encountered is between [this code for MSC2974](https://github.com/matrix-org/matrix-widget-api/blob/b571e1860903b3ef712b62380f042b6ad9b5611f/src/ClientWidgetApi.ts#L241-L248) and the `validateCapabilities` function. While the function was originally designed to only handle the initial capabilities request and no further renegotiations, it still called with the capbilities diff because there is no proper alternative yet. And in the initial negotiation, it would explicitly (and rightfully) forget old approvals if a widgets dropped them from their initial capabilities list.

I wonder why the `ClientWidgetApi` does the filtering to only request the diff between the requested and approved capabilities and just not send the whole list to this function, practically kind-of impersonating the initial negotiation.

Reply:
But with this information, yes, the change should probably be reverted and using an intersection again, until this is fully implemented.

Reply:
I think in practice we'd send not only `requestedCapabilities` but also `approvedCapabilities` so the driver knows what to do with the stuff it already approved, and possibly inform the user's consent a bit.

Reply:
When you talk about the "API surface of the widget-api", you mean [this interface](https://github.com/matrix-org/matrix-widget-api/blob/b571e1860903b3ef712b62380f042b6ad9b5611f/src/driver/WidgetDriver.ts#L39-L53)? So the "Element"/"Host" surfaced part of the widget API and not the widget surfaced part.

Reply:
Yup! That bit would need changing.

Reply:
I updated the PR to revert to the old logic. Is there a ticket that tracks the progress of the planned API change in the widget API? It would be really nice to be able to use the capabilities renegotiation feature in a widget.

Reply:
I've opened https://github.com/matrix-org/matrix-widget-api/issues/52 for now to track it.